compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -452,7 +453,11 @@ public void visit(BLangLiteral literalExpr, AnalyzerData data) {
         if (literalType == symTable.semanticError || literalExpr.isFiniteContext) {
             return;
         }
-        data.resultType = types.checkType(literalExpr, literalType, data.expType);
+        if (literalType.tag == TypeTags.ARRAY) {
+            data.resultType = checkByteArrayType(literalExpr, literalType, data);
+        } else {
+            data.resultType = types.checkType(literalExpr, literalType, data.expType);
+        }
```

Comment:
Why do we need to special case like this? Since the expected type is available when we call `setLiteralValueAndGetType`, isn't it enough to update `setLiteralValueAndGetType` to handle byte array literals when the applicable contextually-expected type is immutable?

Reply:
If we do that according to this logic it will return type mismatch error https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L828-#L831. 

if the contextual expected type is immutable then we check the compatibility of individual member as `list-constructor-expr` logic. Compare to the other literals they are `inherently immutable` but `byte-array-literal` is considered as `array-type` which is `selectively immutable`, that's why we special case that.

Reply:
When it fails, what is `literalType`? Isn't it a subtype of `readonly`? It should be a subtype of `readonly`, right? Which should make checkType pass.

Reply:
Updated 1ea491f08f4d803a466d7804eb8973e3c84b0eb5